connectivity/tests/interruption.go

```diff
@@ -0,0 +1,201 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package tests
+
+import (
+	"context"
+	"encoding/json"
+	gojson "encoding/json"
+	"fmt"
+	"os"
+	"sort"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/cilium-cli/connectivity/check"
+	"github.com/cilium/cilium-cli/defaults"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type interruptionCounts struct {
+	Restarts   map[string]string `json:"restarts"`
+	XfrmErrors map[string]string `json:"xfrm_errors,omitempty"`
+}
+
+type ciliumMetricsXfrmError struct {
+	Labels struct {
+		Error string `json:"error"`
+		Type  string `json:"type"`
+	} `json:"labels"`
+	Name  string `json:"name"`
+	Value uint64 `json:"value"`
+}
+
+// NoInterruptedConnections checks whether there are no interruptions in
+// long-lived E/W LB connections. The test case is used to validate Cilium
+// upgrades, ipsec key rotation, etc.
+//
+// The test case consists of three steps:
+//
+// 1. Deploying pods and a service which establish the long-lived connections
+// (done by "--interruption-test-setup"). The client pods ("test-conn-disrupt-client")
+// establish connections via ClusterIP ("test-conn-disrupt") to server pods
+// ("test-conn-disrupt-server"). As there former pods come first before the latter,
+// the former pods can crash which increases the pod restart counter. The step
+// is responsible for storing the restart counter too.
+// 2. Do Cilium upgrade, or ipsec key rotation, etc.
+// 3. Run the test ("--include-interruption-test"). The test checks the restart
+// counters, and compares them against the previously stored ones. A mismatch
+// indicates that a connection was interrupted.
+func NoInterruptedConnections() check.Scenario {
+	return &noInterruptedConnections{}
+}
+
+type noInterruptedConnections struct{}
+
+func (n *noInterruptedConnections) Name() string {
+	return "no-interrupted-connections"
+}
+
+func (n *noInterruptedConnections) Run(ctx context.Context, t *check.Test) {
+	ct := t.Context()
+	ipsecEnabled := false
+	if encryptionFeature, ok := ct.Feature(check.FeatureEncryptionPod); ok && encryptionFeature.Enabled && encryptionFeature.Mode == "ipsec" {
```

Comment:
I'd prefer to have the XFRM counter checks in a separate `check.Scenario`. Same as we do with the `NoMissedTailCalls`.